### ISSUE TITLE
Support S3 as file storage

### ISSUE THREAD

**Issue by Flowneee** [kellnr/kellnr#173]
Hi. I wanted to ask about S3 support - is it planned? I think it could be very useful in environments where it is available. For example in my current environment I have no access to any kind of persistent storage unfortunately, so deploy in k8s is out of question. For example https://github.com/mcorbin/meuse supports it (we currently use it, but it is lack certain features like docs generation and user management in UI, also written in Clojure so hacking it is much harder).

If you ok with it, I will try to implement it and send a PR.

**Comment by secana** [comment]
Hi,

yes it's planned. When I built kellnr, I abstracted the storage as far as possible, such that a switch of the storage provider is possible. I used the same to support multiple database backends. I don't know when I have time to tackle the S3 backend, but I'll keep this issue as a reminder.

**Comment by MarcAntoine-Arnaud** [comment]
It may relevant to use OpenDAL, so it will abstract various storages, inc. S3 and local and Google Storage etc.

**Comment by secana** [comment]
@MarcAntoine-Arnaud Thanks for the idea. I never heard of OpenDAL before, but it looks very promising.

**Comment by MarcAntoine-Arnaud** [comment]
Yes we use it for an internal project and it's cool. 
We may help to integrate it.

**Comment by Omega359** [comment]
Another solution could be to use [object_store](https://docs.rs/object_store/latest/object_store/).

**Comment by kushudai** [comment]
I took a first hacky stab at this in https://github.com/kellnr/kellnr/pull/600
Feedback is appreciated. It's certainly _not_ ready for a review with the intention of merging that code.

**Comment by secana** [comment]
Hi @kushudai! Thank you for the PR. While it's quite hacky it show that a S3 integration does not seem to be too much work. Are you interested in improving the PR futher?

**Comment by kushudai** [comment]
> Hi [@kushudai](https://github.com/kushudai)! Thank you for the PR. While it's quite hacky it show that a S3 integration does not seem to be too much work. Are you interested in improving the PR futher?

Absolutely, I would appreciate some advice on how we want to deal with the Drop trait requirement on the S3File type :)

**Comment by secana** [comment]
I'll have a look as soon as I've time. I never used S3 from Rust, so that's a new one for me.

**Comment by secana** [comment]
In progress in PR #623
